packages/data/src/components/use-select/index.js

```diff
@@ -19,6 +19,23 @@ import useAsyncMode from '../async-mode-provider/use-async-mode';
 
 const renderQueue = createQueue();
 
+function warnOnUnstableReference( a, b ) {
+	let keys = [];
+	if ( a.constructor === Object && b.constructor === Object ) {
+		keys = Object.keys( a ).filter( ( k ) => a[ k ] !== b[ k ] );
+	} else if ( Array.isArray( a ) && Array.isArray( b ) ) {
+		keys = [ ...a.keys() ].filter( ( i ) => a[ i ] !== b[ i ] );
+	}
```

Comment:
> Object.keys should work equally well also for arrays. The only difference is that .keys() returns numbers, while Object.keys are strings.

Yeah. I should have guessed that 😅 

> The a.constructor check will crash when a is null or undefined, we should guard for that.

I've borrowed the check for `isShallowEqual`, so it should be safe in this case.

Based on your suggestion, I think we can simplify logic to the example below. What do you think?

```js
const keys =
	typeof a === 'object' && typeof b === 'object'
		? Object.keys( a ).filter( ( k ) => a[ k ] !== b[ k ] )
		: [];
```

Reply:
> I've borrowed the check for `isShallowEqual`, so it should be safe in this case.

But `isShallowEqual` does an `if ( a && b )` check before it starts looking at `a.constructor`.

> `typeof a === 'object' && typeof b === 'object'`

`typeof null` is also `'object'` 🙀 It will all work only if you exit early when `a` or `b` is falsy.

Reply:
I should stop ignoring the "everything is an object" JS motto 😄

Pushed update in [39bc8c1](https://github.com/WordPress/gutenberg/pull/67713/commits/39bc8c1c50fd7566f14207b3251c98fb9f1a43b8).